### DISCUSSION TITLE
Can't reason how alternate template/shape works in OrchardCore.Taxonomies?

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#10917] [Q&A] [answered]
I want to override templates of BlogPost's category and tags but I can't reason how it works.
I would be glad if you could explain me for the following questions.

To override an existing `category's template`, this templates work:
- BlogPost-Category.cshtml

However, this template does not work:
- BlogPost-Category-TaxonomyField.cshtml

The questions are:
> Why `BlogPost-Category-TaxonomyField.cshtml` does not match "[PartType]__[FieldName]__[ShapeType]" pattern?
> Is `TaxonomyField` considedered to be both `FieldType` and `ShapeType`(TaxonomyFieldDisplayDriver returns TaxonomyField shape type)?
 
----
To override an existing `tags' template`, only this templates work:
- BlogPost-Tags-TaxonomyField-Tags.Display

The questions are:
> Is this template match `[PartType]__[FieldName]__[ShapeType]` or `[PartType]_[DisplayType]__[FieldName]__[FieldType]_[DisplalyMode]` pattern?
> Why I can't remove `.Display` from the template name to get "BlogPost-Tags-TaxonomyField-Tags.cshtml" work? I don't want to create a specific template because I want to use it in both summary and detail page.

These templates do not work at all:
- BlogPost-Tags.cshtml (It should match [PartType]__[FieldName], the same as BlogPost-Category.cshtml)
- BlogPost-Tags-TaxonomyField.csthml
- BlogPost-Tags-TaxonomyField-Tags.cshtml

The questions is:
> Could you please explain why the above templates do not work?

Thank you so much.

**Comment by ns8482e** [comment, marked as answer]
Category  and Tags in BlogPost are fields not parts, 

Let's deconsturct the template tokens to better understand 

**BlogPost**
[ContentType]  is `BlogPost`
[PartType] is `BlogPost` (implicit part for Category and Tags fields added to content type)
[DisplayType] is `Detail` means it doesn't use `[DisplayType]` is ignred in templates

**Category**
[FieldName] is `Category`
[FieldType] is `TaxonomyField`
[ShapeType] is `TaxonomyField`
[DisplayMode] is `Standard` means it doesn't use `DisplayMode` templates

**Tags**
[FieldName] is `Tags`
[FieldType] is `TaxonomyField`
[ShapeType] is `TaxonomyField`
[DisplayMode] is `Tags` means uses `DisplayMode` templates.


---------------------------------------------------------------------------------


  
> this templates work:
> 
> * BlogPost-Category.cshtml

This works because it maps to `[PartType]__[FieldName]` ( here ContentPart is implicit part `BlogPost` and FieldName is `Category`)




> However, this template does not work:
> 
> * BlogPost-Category-TaxonomyField.cshtml
> 
> The questions are:
> 
> Why `BlogPost-Category-TaxonomyField.cshtml` does not match "[PartType]**[FieldName]**[ShapeType]" pattern?
> Is `TaxonomyField` considedered to be both `FieldType` and `ShapeType`(TaxonomyFieldDisplayDriver returns TaxonomyField shape type)?

This doesn't work because it's doesnt match with the pattern `[PartType]__[FieldName]__[ShapeType]` as `[ShapeType]` is ignored when `[ShapeType]` is same as field type i.e. `TaxonomyField` 

> To override an existing `tags' template`, only this templates work:
> 
> * BlogPost-Tags-TaxonomyField-Tags.Display
> The questions are:
> 
> Is this template match `[PartType]__[FieldName]__[ShapeType]` or `[PartType]_[DisplayType]__[FieldName]__[FieldType]_[DisplalyMode]` pattern?


This works because for Detail it match with the pattern `[PartType]_Display__[FieldName]__[ShapeType]__[DisplayMode]`  which similar to `[PartType]__[FieldName]__[ShapeType]` however with display mode. Here `[ShapeType]` is not ignored due to explicit `[DisplayMode]` 
In case If you are rendering Summary then it would have match `[PartType]_[DisplayType]__[FieldName]__[ShapeType]__[DisplayMode]_Display`

> Why I can't remove `.Display` from the template name to get "BlogPost-Tags-TaxonomyField-Tags.cshtml" work? I don't want to create a specific template because I want to use it in both summary and detail page.

`Display` token is required with display mode. 

> These templates do not work at all:
> 
> * BlogPost-Tags.cshtml (It should match [PartType]__[FieldName], the same as BlogPost-Category.cshtml)
> * BlogPost-Tags-TaxonomyField.csthml
> * BlogPost-Tags-TaxonomyField-Tags.cshtml


If you want Tags to resolve template same as `Category` and want Detail and Summary to resolve same template - change it's display mode from `Tags` to `Standard` in Field settings.

**Reply by kdubious** [reply]
Wonderful. May be useful to create an article from this, @ns8482e ?

**Comment by aaronamm** [comment]
@ns8482e Thank you so much for your prompt reply.
Your answers are very helpful and clear to me.

I just only want to add some information that can be useful for others or reference.


> [ShapeType] is ignored when [ShapeType] is same as field type i.e. TaxonomyField

As the following code:
https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentFieldDisplayDriver.cs#L73

> Display token is required with display mode.

As the following code:
https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentFieldDisplayDriver.cs#L89-L93
and
https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentFieldDisplayDriver.cs#L99-L111

## Get the list of alternate shapes
If we debug and watch a model object in `TaxonomyField.cshtml`.
We can find a list of alternate shapes of Category field from `Model.Metadata.Alternates` property.
This can be  helpful.

![image](https://user-images.githubusercontent.com/2938310/147381960-b3ac5b8e-6f71-4860-9127-a9ae76ac849c.png)

**Comment by Skrypt** [comment]
Using the mini-profiler feature is quite helpful sometimes to figure out which templates are rendered. Also to know which shapes are instanciated.

**Reply by aaronamm** [reply]
@Skrypt Thanks for suggesting Mini Profile.

I would like to make short note of how to use it which can be useful for other people.
- Enable it in admin panel.
![image](https://user-images.githubusercontent.com/2938310/147740636-c7e5cf54-5b5c-42ed-a90b-6771090f920e.png)
- If you want to show Mini Profile on an admin page, add the following configuration to `Startup.cs` in the main CMS project.

```cs
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services
                .AddOrchardCms()
                .AllowMiniProfilerOnAdmin();
        }
    }
```
- Then use it. 
  - Example in an admin page.
  - You may need to toggle show trivial to get more information.
  
![image](https://user-images.githubusercontent.com/2938310/147740853-f8b9eeec-9481-48d2-96d9-f06efc857333.png)


REF
- https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.MiniProfiler/MiniProfilerOrchardCoreBuilderExtensions.cs
- https://docs.orchardcore.net/en/dev/docs/reference/modules/MiniProfiler/

**Comment by aaronamm** [comment]
FYI if we create a custom theme, it need to have a layout zone "footer" to render Mini Profiler.
REF https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.MiniProfiler/MiniProfilerFilter.cs#L37-L42
```cshtml
@* Layout.csthml *@

    <footer>
      @await RenderSectionAsync("Footer", required: false)
    </footer>

```

Example of Mini Profiler on a frontend. 

![image](https://user-images.githubusercontent.com/2938310/147747691-6b9db3be-e699-4603-a584-a90f87dd4e0e.png)
